### PR TITLE
Add support for OpenHub and DeviantArt

### PR DESCRIPTION
**Pull request by seeeturtle**
This will add checkings for OpenHub and DeviantArt.
Uses icon of OpenHub from blackducksoftware/ohloh-ui repo.

Closes https://github.com/manu-chroma/username-availability-checker/issues/24

### CODE REVIEW

**Comment by seeeturtle** [comment]
Do I have to add test for this code?

File: websites.yml [hunk 1]
```diff
@@ -46,3 +48,16 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+  openhub:
```
Comment (jayvdb) [inline websites.yml]: Please re-read https://github.com/blackducksoftware/ohloh-ui/blob/master/app/models/concerns/account_validations.rb again, which I showed to you in Zulip ; there are more rules in there that you have missed. Also log into OpenHub, and change the text in your username setting to find some additional rules which can be added.  (You dont need to save your changes ; it will dynamically tell you about invalid usernames )

File: templates/status.html [hunk 2]
```diff
@@ -47,13 +47,18 @@
...
             // list of supported websites
-            var sites = ["facebook", "twitter", "instagram", "github", "soundcloud", "pinterest", "tumblr", "behance", "gitlab"];
+            var sites = ["facebook", "twitter", "instagram", "github", "soundcloud", "pinterest", "tumblr", "behance", "gitlab", "openhub", "deviantart"];
 
             // create cards dynamically for each of the websites
             sites.forEach(website => {
-              $(".helper").append(`<div class="card">
-                <p><div class="tooltip"><i class="fa fa-${website}"></i><span class="tooltiptext">${website}</span></div>
-                <span id='${website}'><i class="fa fa-circle-o-notch fa-spin"></i></span></p>`)
+              if (website != "openhub")
```
Comment (jayvdb) [inline templates/status.html]: please dont change this code.  Instead, use CSS rules to load the OpenHub logo.  It will have a class `fa-openhub` that you can use.

File: websites.yml (hunk 1 above)
Comment (seeeturtle) [reply websites.yml]: sorry, I forgot to do it. but openhub doesn't dynamically tell me :(

File: templates/status.html (hunk 2 above)
Comment (seeeturtle) [reply templates/status.html]: ok

Comment (jayvdb) [reply templates/status.html]: this is still here.

File: websites.yml (hunk 1 above)
Comment (jayvdb) [reply websites.yml]: still not done.  Please re-read https://github.com/blackducksoftware/ohloh-ui/blob/master/app/models/concerns/account_validations.rb carefully

Comment (jayvdb) [reply websites.yml]: still not done.  no guessing required.  Just read that code.

**Comment by manu-chroma** [comment]
> Do I have to add test for this code?

Yes. Refer to the existing tests. Would just be a matter of replicating the same code

**Comment by manu-chroma** [comment]
@seeeturtle Please rebase before proceeding further with your work.

File: websites.yml (hunk 1 above)
Comment (jayvdb) [reply websites.yml]: Looks better now

File: username_api.py [hunk 3]
```diff
@@ -73,10 +73,14 @@ def check_username(website, username):
 			'possible': possible,
 		}
 
-	if website in ['pinterest', 'gitlab']:
+	if website in ['pinterest', 'gitlab', 'openhub']:
 		res  = r.get(url)
-		code = 200 if bytes(username, encoding='utf-8') in res.content \
-			else 404
+		if website == 'openhub':
```
Comment (jayvdb) [inline username_api.py]: why do you need to do this?  It returns 404

```
wget -S https://www.openhub.net/accounts/jayvdb123
--2017-12-23 13:25:35--  https://www.openhub.net/accounts/jayvdb123
Resolving www.openhub.net (www.openhub.net)... 192.231.134.139
Connecting to www.openhub.net (www.openhub.net)|192.231.134.139|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 404 Not Found
```

File: tests/test_username_api.py [hunk 4]
```diff
@@ -13,9 +13,16 @@
 
 invalid_username = '$very%long{invalid}user(name)'
 
-def generate_random_username(length):
-  return ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase \
-    + string.digits) for _ in range(length))
+def generate_random_username(length, first_char_letter=False):
+  return ''.join(generate_random_character(length, first_char_letter))
+
+def generate_random_character(length, first_char_letter):
```
Comment (jayvdb) [inline tests/test_username_api.py]: this should be a separate commit. It is being applied to all of the websites, which is good.

File: websites.yml (hunk 1 above)
Comment (seeeturtle) [reply websites.yml]: thanks.

File: tests/test_username_api.py (hunk 4 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: what do you mean "It is being applied to all of the websites"?

File: username_api.py (hunk 3 above)
Comment (seeeturtle) [reply username_api.py]: what? It responded 200 when I try. ok

File: tests/test_username_api.py (hunk 4 above)
Comment (jayvdb) [reply tests/test_username_api.py]: This change, is a separate enhancement, and it is enhancing all sites, not just OpenHub & DeviantArt.
Therefore, it should be (at least) a separate commit in this PR.

You could also do a separate issue & PR for it, as it is not necessary for the rest of your PR.  I can even create a GCI task for it.

Comment (seeeturtle) [reply tests/test_username_api.py]: got it. But can I extract part of changes of a commit?

Comment (seeeturtle) [reply tests/test_username_api.py]: nevermind. I found solution.

File: websites.yml [hunk 5]
```diff
@@ -58,3 +73,5 @@ avatar:
   tumblr: opengraph
   twitter:
     url: https://twitter.com/{u}/profile_image?size=original
+  openhub: false
```
Comment (jayvdb) [inline websites.yml]: order these keys alphabetically

Comment (seeeturtle) [reply websites.yml]: sure

File: websites.yml [hunk 6]
```diff
@@ -46,6 +48,19 @@ username_patterns:
     invalid_patterns:
       - "\\.(com|net)"
       - "(\\.)\\1{1,}"  # consecutive '.' not allowed
+  openhub:
+    characters: a-zA-Z0-9_-
+    min_length: 3
+    max_length: 40
+    invalid_patterns:
+      - "^[^a-zA-Z]"
+  deviantart:
```
Comment (jayvdb) [inline websites.yml]: also add these keys in alpha order, but dont re-order the existing ones.

i.e. do 'best effort' or make your changes roughly ordered correctly.  Later we'll re-order the other ones.

File: websites.yml [hunk 7]
```diff
@@ -4,6 +4,8 @@ urls:
   pinterest: https://in.{w}.com/{u}
   tumblr: https://{u}.{w}.com
   facebook: https://mbasic.{w}.com/{u}
+  openhub: https://www.{w}.net/accounts/{u}
+  deviantart: https://{u}.{w}.com/
```
Comment (jayvdb) [inline websites.yml]: again, best effort alpha order

File: websites.yml (hunk 6 above)
Comment (seeeturtle) [reply websites.yml]: hmm... what's 'best effort'? what do you mean "make your changes roughly ordered correctly"? @jayvdb

Comment (seeeturtle) [reply websites.yml]: and what key should I sort?

Comment (jayvdb) [reply websites.yml]: `deviantart` should be after `behance`

**Comment by seeeturtle** [comment]
got it ;-)

2017. 12. 23. 오후 4:37 John Vandenberg <notifications@github.com> 작성:

File: websites.yml (hunk 6 above)
Comment (seeeturtle) [reply websites.yml]: got it ;-)

File: tests/test_data.yml [hunk 8]
```diff
@@ -57,3 +57,13 @@ pinterest:
     - "steve"
   available_usernames:
     - "zNRe3jx3isA8CoM"
+openhub:
+  taken_usernames:
```
Comment (manu-chroma) [inline tests/test_data.yml]: can you add 2 more taken usernames for both these websites?

**Comment by manu-chroma** [comment]
Change this line in commit: 
``It's made because of to suppport OpenHurb.``
to something like
``Used in OpenHub testing``

File: tests/test_data.yml (hunk 8 above)
Comment (seeeturtle) [reply tests/test_data.yml]: sure

**Comment by manu-chroma** [comment]
@jayvdb any other comments?

File: tests/test_data.yml [hunk 9]
```diff
@@ -57,3 +57,17 @@ pinterest:
...
+    - "seeeturtle"
+    - "andrewda"
+    - "nibnalin"
+  available_usernames:
+    - "zNRe3jx3isA8CoM"
+deviantart:
```
Comment (jayvdb) [inline tests/test_data.yml]: insert in roughly alpha order; i.e. `deviantart` at the beginning. `openhub` before `pinterest`
